File: cmd/top.go
package cmd

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/abix-/k3sc/internal/format"
	"github.com/abix-/k3sc/internal/github"
	"github.com/abix-/k3sc/internal/k8s"
	"github.com/abix-/k3sc/internal/tui"
	"github.com/abix-/k3sc/internal/types"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/spf13/cobra"
	"k8s.io/client-go/kubernetes"
	"k8s.io/klog/v2"
)

// ringBuffer captures the last N lines written to it.
type ringBuffer struct {
	mu    sync.Mutex
	lines []string
	max   int
	buf   bytes.Buffer
}

func (r *ringBuffer) Write(p []byte) (int, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.buf.Write(p)
	for {
		line, err := r.buf.ReadString('\n')
		if err != nil {
			// incomplete line, put it back
			r.buf.WriteString(line)
			break
		}
		line = strings.TrimRight(line, "\r\n")
		if line != "" {
			r.lines = append(r.lines, line)
			if len(r.lines) > r.max {
				r.lines = r.lines[1:]
			}
		}
	}
	return len(p), nil
}

func (r *ringBuffer) Lines() []string {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]string, len(r.lines))
	copy(out, r.lines)
	return out
}

var once bool

func init() {
	topCmd.Flags().BoolVar(&once, "once", false, "Print once and exit (no TUI)")
	rootCmd.AddCommand(topCmd)
}

var topCmd = &cobra.Command{
	Use:   "top",
	Short: "Dashboard of agent pods, GitHub issues, and cluster health",
	RunE:  runTop,
}

type dashboard struct {
	nodeName    string
	nodeVersion string
	dispatch    types.DispatchStateInfo
	pods        []types.AgentPod
	tasks       []types.TaskInfo
	issues      []types.Issue
	prs         []types.PullRequest
	operatorLog string
	timberbot   types.TimberbotInfo
}

func gather(cs *kubernetes.Clientset) (*dashboard, error) {
	ctx := context.Background()

	var (
		nodeName, nodeVersion string
		dispatchState         types.DispatchStateInfo
		pods                  []types.AgentPod
		tasks                 []types.TaskInfo
		issues                []types.Issue
		prs                   []types.PullRequest
		dispLog               string
		timberbot             types.TimberbotInfo
		mu                    sync.Mutex
		wg                    sync.WaitGroup
		errs                  []error
	)

	wg.Add(8)
	go func() {
		defer wg.Done()
		n, v, e := k8s.GetNodeInfo(ctx, cs)
		mu.Lock()
		nodeName, nodeVersion = n, v
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	go func() {
		defer wg.Done()
		s, e := k8s.GetDispatchState(ctx)
		mu.Lock()
		dispatchState = s
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	go func() {
		defer wg.Done()
		p, e := k8s.GetAgentPods(ctx, cs)
		mu.Lock()
		pods = p
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	go func() {
		defer wg.Done()
		t, e := k8s.GetAgentJobs(ctx)
		mu.Lock()
		tasks = t
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	go func() {
		defer wg.Done()
		i, e := github.GetWorkflowIssues(ctx)
		mu.Lock()
		issues = i
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	go func() {
		defer wg.Done()
		p, e := github.GetOpenPRs(ctx)
		mu.Lock()
		prs = p
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	go func() {
		defer wg.Done()
		d, e := k8s.GetOperatorLog(ctx, cs)
		mu.Lock()
		dispLog = d
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	go func() {
		defer wg.Done()
		t, e := k8s.GetTimberbotSpec(ctx)
		mu.Lock()
		timberbot = t
		if e != nil {
			errs = append(errs, e)
		}
		mu.Unlock()
	}()
	wg.Wait()

	return &dashboard{
		nodeName:    nodeName,
		nodeVersion: nodeVersion,
		dispatch:    dispatchState,
		pods:        pods,
		tasks:       tasks,
		issues:      issues,
		prs:         prs,
		operatorLog: dispLog,
		timberbot:   timberbot,
	}, nil
}

func hydratePodLogTails(ctx context.Context, cs *kubernetes.Clientset, pods []types.AgentPod) {
	var wg sync.WaitGroup
	for i := range pods {
		if pods[i].Phase != types.PhaseRunning && pods[i].Phase != types.PhasePending {
			continue
		}
		wg.Add(1)
		go func(idx int) {
			defer wg.Done()
			tail, _ := k8s.GetPodLogTail(ctx, cs, pods[idx].Name, 20)
			pods[idx].LogTail = tail
		}(i)
	}
	wg.Wait()
}

func applyLiveLogTails(pods []types.AgentPod, liveLogs []tui.LiveLog) {
	tailsByPod := make(map[string]string, len(liveLogs))
	for _, log := range liveLogs {
		tailsByPod[log.PodName] = log.Tail
	}
	for i := range pods {
		if tail, ok := tailsByPod[pods[i].Name]; ok {
			pods[i].LogTail = tail
		} else if pods[i].Phase == types.PhaseRunning {
			pods[i].LogTail = ""
		}
	}
}

func mergeTaskRuntime(tasks []types.TaskInfo, pods []types.AgentPod) []types.TaskInfo {
	if len(tasks) == 0 {
		return nil
	}

	latestByKey := make(map[string]types.AgentPod, len(pods))
	for _, pod := range pods {
		key := taskPodKey(pod.Repo, pod.Issue, types.AgentName(pod.Family, pod.Slot))
		if current, ok := latestByKey[key]; !ok || newerPod(pod, current) {
			latestByKey[key] = pod
		}
	}

	merged := make([]types.TaskInfo, len(tasks))
	copy(merged, tasks)
	for i := range merged {
		if pod, ok := latestByKey[taskPodKey(merged[i].Repo, merged[i].Issue, merged[i].Agent)]; ok {
			merged[i].RuntimePhase = pod.Phase
			merged[i].LogTail = pod.LogTail
		}
	}

	sort.SliceStable(merged, func(i, j int) bool {
		if merged[i].PhaseOrder() != merged[j].PhaseOrder() {
			return merged[i].PhaseOrder() < merged[j].PhaseOrder()
		}
		ti, tj := taskEventTime(merged[i]), taskEventTime(merged[j])
		switch {
		case ti.Equal(tj):
			return merged[i].Issue > merged[j].Issue
		case ti.IsZero():
			return false
		case tj.IsZero():
			return true
		default:
			return ti.After(tj)
		}
	})

	return merged
}

func taskPodKey(repo types.Repo, issue int, agent string) string {
	return strings.ToLower(fmt.Sprintf("%s/%s#%d:%s", repo.Owner, repo.Name, issue, agent))
}

func newerPod(a, b types.AgentPod) bool {
	ta, tb := podEventTime(a), podEventTime(b)
	switch {
	case ta.Equal(tb):
		return a.Name > b.Name
	case ta.IsZero():
		return false
	case tb.IsZero():
		return true
	default:
		return ta.After(tb)
	}
}

func podEventTime(p types.AgentPod) time.Time {
	if p.Finished != nil {
		return *p.Finished
	}
	if p.Started != nil {
		return *p.Started
	}
	return time.Time{}
}

func taskEventTime(t types.TaskInfo) time.Time {
	if t.Started != nil {
		return *t.Started
	}
	if t.Finished != nil {
		return *t.Finished
	}
	return time.Time{}
}

func countTaskPhases(tasks []types.TaskInfo) (running, done, failed, blocked int) {
	for _, t := range tasks {
		switch t.Phase {
		case "Running", "Pending":
			running++
		case "Succeeded":
			done++
		case "Failed":
			failed++
		case "Blocked":
			blocked++
		}
	}
	return running, done, failed, blocked
}

func runtimeLabel(phase types.PodPhase) string {
	if phase == "" {
		return "-"
	}
	return phase.Display()
}

func printDashboard(d *dashboard) {
	fmt.Println("=== CLUSTER ===")
	fmt.Printf("Node: %s Ready %s\n\n", d.nodeName, d.nodeVersion)

	// 1. Operator
	fmt.Println("=== QUOTA ===")
	if len(d.dispatch.FamilyStatuses) == 0 {
		fmt.Println("  (quota unknown)")
	} else {
		for _, status := range d.dispatch.FamilyStatuses {
			label := "unknown"
			if status.Checked && status.Available {
				label = "ok"
			} else if !status.Available {
				label = "blocked"
			}
			reason := status.Reason
			if reason != "" {
				fmt.Printf("  %-6s %-8s %s\n", status.Family, label, reason)
			} else {
				fmt.Printf("  %-6s %s\n", status.Family, label)
			}
		}
	}
	fmt.Println()

	// timberbot
	if d.timberbot.Enabled {
		fmt.Println("=== TIMBERBOT ===")
		fmt.Printf("  status: enabled\n")
		fmt.Printf("  goal: %s\n", d.timberbot.Goal)
		fmt.Printf("  rounds: %d\n", d.timberbot.Rounds)
		var tbPod *types.AgentPod
		for i := range d.pods {
			if d.pods[i].JobKind == "timberbot" && (d.pods[i].Phase == types.PhaseRunning || d.pods[i].Phase == types.PhasePending) {
				tbPod = &d.pods[i]
				break
			}
		}
		if tbPod != nil {
			agent := types.AgentName(tbPod.Family, tbPod.Slot)
			fmt.Printf("  agent: %s  pod: %s  uptime: %s\n", agent, tbPod.Name, format.FmtDuration(tbPod.Started, nil))
		} else {
			fmt.Printf("  agent: (waiting for dispatch)\n")
		}
		fmt.Println()
	}

	// 2. Local review
	fmt.Println("=== LOCAL REVIEW ===")
	if len(d.dispatch.ReviewReservations) == 0 {
		fmt.Println("  (no local PR reservations)")
	} else {
		fmt.Printf("%-12s %-12s %-7s %-8s %-16s %-16s Branch\n", "Worker", "Repo", "PR", "Issue", "Reserved", "Expires")
		for _, res := range d.dispatch.ReviewReservations {
			issue := ""
			if res.Issue > 0 {
				issue = fmt.Sprintf("#%d", res.Issue)
			}
			fmt.Printf("%-12s %-12s %-7s %-8s %-16s %-16s %s\n",
				res.WorkerID,
				res.Repo.Name,
				format.PRLink(res.Repo, res.PRNumber),
				issue,
				format.FmtTime(res.ReservedAt),
				format.FmtTime(res.ExpiresAt),
				res.Branch)
		}
	}
	fmt.Println()

	// 3. Operator
	fmt.Println("=== OPERATOR ===")
	if d.operatorLog == "" {
		fmt.Println("  (no operator logs)")
	} else {
		for _, line := range strings.Split(strings.TrimSpace(d.operatorLog), "\n") {
			if line != "" {
				fmt.Printf("  %s\n", line)
			}
		}
	}
	fmt.Println()

	// 4. Issues
	fmt.Println("=== GITHUB ISSUES ===")
	if len(d.issues) == 0 {
		fmt.Println("  (no issues with workflow labels)")
	} else {
		fmt.Printf("%-7s %-12s %-14s %-10s Title\n", "Issue", "Repo", "State", "Owner")
		for _, i := range d.issues {
			fmt.Printf("%s %-12s %-14s %-10s %s\n", format.IssueLink(i.Repo, i.Number), i.Repo.Name, i.State, i.Owner, i.Title)
		}
	}
	fmt.Println()

	// 5. Operator Tasks
	running, done, failed, blocked := countTaskPhases(d.tasks)
	fmt.Printf("=== OPERATOR TASKS (%d running, %d done, %d failed, %d blocked) ===\n", running, done, failed, blocked)
	if len(d.tasks) == 0 {
		fmt.Println("  (no operator tasks)")
	} else {
		fmt.Printf("%-7s %-10s %-10s %-10s %-10s %-16s %-10s %-13s %s\n", "Issue", "Repo", "Agent", "Task", "Runtime", "Started", "Duration", "Next", "Last Output")
		for _, task := range d.tasks {
			fmt.Printf("%s %-10s %-10s %-10s %-10s %-16s %-10s %-13s %s\n",
				format.IssueLink(task.Repo, task.Issue),
				task.Repo.Name,
				task.Agent,
				task.Phase,
				runtimeLabel(task.RuntimePhase),
				format.FmtTime(task.Started),
				format.FmtDuration(task.Started, task.Finished),
				task.NextAction,
				task.LogTail)
		}
	}
	fmt.Println()

	fmt.Println("=== PULL REQUESTS ===")
	if len(d.prs) == 0 {
		fmt.Println("  (no open pull requests)")
	} else {
		fmt.Printf("%-7s %-12s %-12s %-20s Title\n", "PR", "Repo", "Owner", "Branch")
		for _, pr := range d.prs {
			fmt.Printf("%s %-12s %-12s %-20s %s\n", format.PRLink(pr.Repo, pr.Number), pr.Repo.Name, pr.Owner, pr.Branch, pr.Title)
		}
	}
	fmt.Println()
}

func runTop(cmd *cobra.Command, args []string) error {
	cs, err := k8s.NewClient()
	if err != nil {
		return err
	}

	if once {
		d, err := gather(cs)
		if err != nil {
			return err
		}
		hydratePodLogTails(context.Background(), cs, d.pods)
		d.tasks = mergeTaskRuntime(d.tasks, d.pods)
		printDashboard(d)
		return nil
	}

	// capture klog output to ring buffer instead of stderr
	logBuf := &ringBuffer{max: 50}
	klog.SetOutput(logBuf)
	klog.LogToStderr(false)

	// TUI mode
	streamer := tui.NewLogStreamer(cs, types.Namespace)
	defer streamer.Stop()

	gatherFn := func() (*tui.Data, error) {
		d, err := gather(cs)
		if err != nil {
			return nil, err
		}
		streamer.Sync(d.pods)
		liveLogs := streamer.Snapshot()
		applyLiveLogTails(d.pods, liveLogs)
		tasks := mergeTaskRuntime(d.tasks, d.pods)
		return &tui.Data{
			NodeName:    d.nodeName,
			NodeVersion: d.nodeVersion,
			Dispatch:    d.dispatch,
			Pods:        d.pods,
			Tasks:       tasks,
			Issues:      d.issues,
			PRs:         d.prs,
			OperatorLog: d.operatorLog,
			LiveLogs:    liveLogs,
			Timberbot:   d.timberbot,
		}, nil
	}

	k8sGatherFn := func(current *tui.Data) (*tui.Data, error) {
		ctx := context.Background()

		var (
			dispatch  types.DispatchStateInfo
			pods      []types.AgentPod
			dispLog   string
			tbInfo    types.TimberbotInfo
			wg2       sync.WaitGroup
			mu2       sync.Mutex
		)
		wg2.Add(4)
		go func() {
			defer wg2.Done()
			s, _ := k8s.GetDispatchState(ctx)
			mu2.Lock()
			dispatch = s
			mu2.Unlock()
		}()
		go func() {
			defer wg2.Done()
			p, _ := k8s.GetAgentPods(ctx, cs)
			mu2.Lock()
			pods = p
			mu2.Unlock()
		}()
		go func() {
			defer wg2.Done()
			d, _ := k8s.GetOperatorLog(ctx, cs)
			mu2.Lock()
			dispLog = d
			mu2.Unlock()
		}()
		go func() {
			defer wg2.Done()
			t, _ := k8s.GetTimberbotSpec(ctx)
			mu2.Lock()
			tbInfo = t
			mu2.Unlock()
		}()
		wg2.Wait()

		// sync streaming logs
		streamer.Sync(pods)
		liveLogs := streamer.Snapshot()
		applyLiveLogTails(pods, liveLogs)
		tasks, _ := k8s.GetAgentJobs(ctx)
		tasks = mergeTaskRuntime(tasks, pods)

		return &tui.Data{
			NodeName:    current.NodeName,
			NodeVersion: current.NodeVersion,
			Dispatch:    dispatch,
			Pods:        pods,
			Tasks:       tasks,
			Issues:      current.Issues,
			PRs:         current.PRs,
			OperatorLog: dispLog,
			LiveLogs:    liveLogs,
			Timberbot:   tbInfo,
		}, nil
	}

	dispatchFn := func() (string, error) {
		return k8s.TriggerDispatch(context.Background())
	}

	maxSlots := 5
	setMaxSlots := func(n int) {
		maxSlots = n
		os.Setenv("MAX_SLOTS", fmt.Sprintf("%d", n))
	}

	m := tui.NewModel(gatherFn, k8sGatherFn, dispatchFn, maxSlots, setMaxSlots, logBuf.Lines)
	p := tea.NewProgram(m, tea.WithAltScreen())
	_, err = p.Run()
	return err
}

File: src/top.rs
use anyhow::Result;
use chrono::{DateTime, Utc};
use chrono_tz::America::New_York;

use crate::github;
use crate::k8s;
use crate::types::{AgentPod, Issue, PodPhase, SLOT_OFFSET};

pub fn fmt_time(dt: DateTime<Utc>) -> String {
    let et = dt.with_timezone(&New_York);
    et.format("%-I:%M %p %Z").to_string()
}

pub fn fmt_duration(start: DateTime<Utc>, end: Option<DateTime<Utc>>) -> String {
    let end = end.unwrap_or_else(Utc::now);
    let secs = (end - start).num_seconds().max(0);
    format!("{}m {:02}s", secs / 60, secs % 60)
}

struct Dashboard {
    node_name: String,
    node_version: String,
    pods: Vec<AgentPod>,
    issues: Vec<Issue>,
    dispatcher_log: String,
}

async fn gather() -> Result<Dashboard> {
    let client = k8s::new_client().await?;

    // parallel fetch
    let (node_info, mut pods, issues, dispatcher_log) = tokio::try_join!(
        k8s::get_node_info(&client),
        k8s::get_agent_pods(&client),
        github::get_workflow_issues(),
        k8s::get_dispatcher_log(&client),
    )?;

    // fetch log tails in parallel
    let mut handles = Vec::new();
    for pod in &pods {
        let c = client.clone();
        let name = pod.name.clone();
        handles.push(tokio::spawn(async move {
            k8s::get_pod_log_tail(&c, &name, 20).await.unwrap_or_default()
        }));
    }
    for (i, handle) in handles.into_iter().enumerate() {
        pods[i].log_tail = handle.await.unwrap_or_default();
    }

    // sort: running first, then completed, then failed -- oldest first within each group
    pods.sort_by(|a, b| a.phase.cmp(&b.phase).then(a.started.cmp(&b.started)));

    Ok(Dashboard {
        node_name: node_info.0,
        node_version: node_info.1,
        pods,
        issues,
        dispatcher_log,
    })
}

fn print_dashboard(d: &Dashboard) {
    println!("=== CLUSTER ===");
    println!("Node: {} Ready {}", d.node_name, d.node_version);
    println!();

    // 1. Issues first
    println!("=== GITHUB ISSUES ===");
    if d.issues.is_empty() {
        println!("  (no issues with workflow labels)");
    } else {
        println!("{:<7} {:<14} {:<10} Title", "Issue", "State", "Owner");
        for i in &d.issues {
            println!("#{:<6} {:<14} {:<10} {}", i.number, i.state, i.owner, i.title);
        }
    }
    println!();

    // 2. Dispatcher
    println!("=== DISPATCHER ===");
    if d.dispatcher_log.is_empty() {
        println!("  (no dispatcher runs found)");
    } else {
        for line in d.dispatcher_log.lines() {
            println!("  {line}");
        }
    }
    println!();

    // 3. Agents last
    let running = d.pods.iter().filter(|p| p.phase == PodPhase::Running).count();
    let completed = d.pods.iter().filter(|p| p.phase == PodPhase::Succeeded).count();
    let failed = d.pods.iter().filter(|p| p.phase == PodPhase::Failed).count();

    println!("=== AGENTS ({running} running, {completed} completed, {failed} failed) ===");
    if d.pods.is_empty() {
        println!("  (no agent pods)");
    } else {
        println!(
            "{:<7} {:<10} {:<11} {:<16} {:<10} Last Output",
            "Issue", "Agent", "Status", "Started", "Duration"
        );
        for pod in &d.pods {
            let agent = format!("claude-{}", pod.slot as u16 + SLOT_OFFSET as u16);
            let started = pod.started.map(fmt_time).unwrap_or_default();
            let duration = pod.started.map(|s| fmt_duration(s, pod.finished)).unwrap_or_default();
            println!(
                "#{:<6} {:<10} {:<11} {:<16} {:<10} {}",
                pod.issue,
                agent,
                pod.phase.display(),
                started,
                duration,
                pod.log_tail
            );
        }
    }
    println!();
}

pub async fn run(once: bool) -> Result<()> {
    if once {
        let d = gather().await?;
        print_dashboard(&d);
        return Ok(());
    }

    // TUI mode with ratatui
    run_tui().await
}

async fn run_tui() -> Result<()> {
    use crossterm::event::{self, Event, KeyCode, KeyEventKind};
    use crossterm::terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen};
    use crossterm::execute;
    use ratatui::prelude::*;
    use ratatui::widgets::*;
    use std::io;
    use std::time::Duration;

    enable_raw_mode()?;
    let mut stdout = io::stdout();
    execute!(stdout, EnterAlternateScreen)?;
    let backend = CrosstermBackend::new(stdout);
    let mut terminal = Terminal::new(backend)?;

    let mut dashboard = gather().await?;
    let mut last_refresh = std::time::Instant::now();
    let mut status_msg = String::new();

    loop {
        // refresh every 5 seconds
        if last_refresh.elapsed() > Duration::from_secs(5) {
            if let Ok(d) = gather().await {
                dashboard = d;
            }
            last_refresh = std::time::Instant::now();
            if !status_msg.is_empty() {
                status_msg = String::new();
            }
        }

        terminal.draw(|f| {
            let chunks = Layout::default()
                .direction(Direction::Vertical)
                .constraints([
                    Constraint::Length(3),  // cluster
                    Constraint::Length(12), // issues
                    Constraint::Length(6),  // dispatcher
                    Constraint::Min(6),    // agents
                    Constraint::Length(1),  // help
                ])
                .split(f.area());

            // cluster
            let running = dashboard.pods.iter().filter(|p| p.phase == PodPhase::Running).count();
            let completed = dashboard.pods.iter().filter(|p| p.phase == PodPhase::Succeeded).count();
            let cluster = Paragraph::new(format!(
                " Node: {} {}  |  Agents: {} running, {} completed",
                dashboard.node_name, dashboard.node_version, running, completed
            ))
            .block(Block::default().borders(Borders::ALL).title(" Cluster "));
            f.render_widget(cluster, chunks[0]);

            // 1. issues
            let issue_header = Row::new(["Issue", "State", "Owner", "Title"])
                .style(Style::default().bold());
            let issue_rows: Vec<Row> = dashboard.issues.iter().map(|i| {
                let style = match i.state.as_str() {
                    "claimed" => Style::default().fg(Color::Yellow),
                    "needs-human" => Style::default().fg(Color::Magenta),
                    "needs-review" => Style::default().fg(Color::Cyan),
                    "ready" => Style::default().fg(Color::Green),
                    _ => Style::default(),
                };
                Row::new([
                    format!("#{}", i.number),
                    i.state.clone(),
                    i.owner.clone(),
                    i.title.clone(),
                ]).style(style)
            }).collect();
            let issue_table = Table::new(issue_rows, [
                Constraint::Length(7),
                Constraint::Length(14),
                Constraint::Length(10),
                Constraint::Fill(1),
            ])
            .header(issue_header)
            .block(Block::default().borders(Borders::ALL).title(" GitHub Issues "));
            f.render_widget(issue_table, chunks[1]);

            // 2. dispatcher log
            let disp_lines: String = dashboard.dispatcher_log
                .lines()
                .map(|l| format!(" {l}"))
                .collect::<Vec<_>>()
                .join("\n");
            let dispatcher = Paragraph::new(disp_lines)
                .block(Block::default().borders(Borders::ALL).title(" Dispatcher (last run) "))
                .style(Style::default().fg(Color::DarkGray));
            f.render_widget(dispatcher, chunks[2]);

            // 3. agents table
            let header = Row::new(["Issue", "Agent", "Status", "Started", "Duration", "Last Output"])
                .style(Style::default().bold());
            let rows: Vec<Row> = dashboard.pods.iter().map(|pod| {
                let agent = format!("claude-{}", pod.slot as u16 + SLOT_OFFSET as u16);
                let started = pod.started.map(fmt_time).unwrap_or_default();
                let duration = pod.started.map(|s| fmt_duration(s, pod.finished)).unwrap_or_default();
                let tail = pod.log_tail.clone();
                let style = match pod.phase {
                    PodPhase::Running => Style::default().fg(Color::Green),
                    PodPhase::Failed => Style::default().fg(Color::Red),
                    _ => Style::default().fg(Color::DarkGray),
                };
                Row::new([
                    format!("#{}", pod.issue),
                    agent,
                    pod.phase.display().to_string(),
                    started,
                    duration,
                    tail,
                ]).style(style)
            }).collect();
            let table = Table::new(rows, [
                Constraint::Length(7),
                Constraint::Length(10),
                Constraint::Length(11),
                Constraint::Length(16),
                Constraint::Length(10),
                Constraint::Fill(1),
            ])
            .header(header)
            .block(Block::default().borders(Borders::ALL).title(" Agents "));
            f.render_widget(table, chunks[3]);

            // help bar + status
            let help_text = match status_msg.as_str() {
                "" => " q: quit  |  n: dispatch now  |  r: refresh  |  refreshes every 5s".to_string(),
                msg => format!(" {msg}"),
            };
            let help_style = if status_msg.is_empty() {
                Style::default().fg(Color::DarkGray)
            } else {
                Style::default().fg(Color::Yellow)
            };
            let help = Paragraph::new(help_text).style(help_style);
            f.render_widget(help, chunks[4]);
        })?;

        // handle input
        if event::poll(Duration::from_millis(250))? {
            if let Event::Key(key) = event::read()? {
                if key.kind == KeyEventKind::Press {
                    match key.code {
                        KeyCode::Char('q') => break,
                        KeyCode::Char('n') => {
                            status_msg = "dispatching...".to_string();
                            match crate::dispatch::run().await {
                                Ok(log) => {
                                    dashboard.dispatcher_log = log;
                                    status_msg = "dispatch complete".to_string();
                                }
                                Err(e) => status_msg = format!("dispatch error: {e}"),
                            }
                            // refresh pods + issues to show new state
                            if let Ok(d) = gather().await {
                                // keep the fresh dispatch log, update everything else
                                let disp_log = dashboard.dispatcher_log.clone();
                                dashboard = d;
                                dashboard.dispatcher_log = disp_log;
                            }
                            last_refresh = std::time::Instant::now();
                        }
                        KeyCode::Char('r') => {
                            if let Ok(d) = gather().await {
                                dashboard = d;
                            }
                            last_refresh = std::time::Instant::now();
                            status_msg = String::new();
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    disable_raw_mode()?;
    execute!(terminal.backend_mut(), LeaveAlternateScreen)?;
    Ok(())
}
